### PR TITLE
(buddy) Github Actions CI

### PR DESCRIPTION
**Pull request by r0mant**
Buddy PR for https://github.com/gravitational/teleport/pull/18211 with some tweaks e.g.:

- Removed GCP auth and logs upload as we've no use for them.
- Renamed workflows for better consistency.
- Large runners for CPU intensive workflows.

### CODE REVIEW

File: .github/workflows/integration-tests-non-root.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      - '**.go'
+      - 'go.mod'
+      - 'go.sum'
+
+jobs:
+  test:
+    name: Integration tests (non-root)
+    runs-on: ubuntu-latest
+
+    permissions:
```
Comment (reedloden) [inline .github/workflows/integration-tests-non-root.yaml:19]: You also need `packages: read` here

File: .github/workflows/unit-tests-code.yaml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      - '**.go'
+      - 'go.mod'
+      - 'go.sum'
+
+jobs:
+  test:
+    name: Unit tests (code)
+    runs-on: ubuntu-latest
+
+    permissions:
```
Comment (reedloden) [inline .github/workflows/unit-tests-code.yaml:19]: You need `packages: read` here

**Review by russjones — APPROVED** [review]

**Review by jakule — APPROVED** [review]

File: .github/workflows/integration-tests-non-root.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      - 'go.sum'
+
+jobs:
+  test:
+    name: Integration Tests (Non-root)
+    runs-on: ubuntu-22.04-16core
+
+    permissions:
+      contents: read
+      id-token: write
```
Comment (reedloden) [inline .github/workflows/integration-tests-non-root.yaml:21]: Is this needed here?

File: .github/workflows/integration-tests-root.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      - 'go.sum'
+
+jobs:
+  test:
+    name: Integration Tests (Root)
+    runs-on: ubuntu-22.04-16core
+
+    permissions:
+      contents: read
+      id-token: write
```
Comment (reedloden) [inline .github/workflows/integration-tests-root.yaml:21]: Is this needed?

File: .github/workflows/lint.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+name: Lint (Go)
+run-name: make lint
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+jobs:
+  lint:
```
Comment (reedloden) [inline .github/workflows/lint.yaml:10]: This is missing `permissions`.

File: .github/workflows/os-compatibility-test.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
+name: OS Compatibility Test
+run-name: OS Compatibility Test
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+jobs:
+  build:
```
Comment (reedloden) [inline .github/workflows/os-compatibility-test.yaml:10]: This is missing `permissions`

File: .github/workflows/os-compatibility-test.yaml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        run: |
+          make build/tctl build/tsh build/tbot build/teleport
+
+      - name: Upload binaries
+        uses: actions/upload-artifact@v3
+        with:
+          name: build
+          path: ${{ github.workspace }}/build/
+
+  test-compat:
```
Comment (reedloden) [inline .github/workflows/os-compatibility-test.yaml:35]: This is missing `permissions`

File: .github/workflows/unit-tests-code-bypass.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+name: Unit Tests (Go)
+run-name: Unit Tests (Go) - ${{ github.run_id }} - @${{ github.actor }}
+
+on:
+  pull_request:
+    paths-ignore:
+      - '**.go'
+
+jobs:
+  test:
```
Comment (reedloden) [inline .github/workflows/unit-tests-code-bypass.yaml:10]: This is missing `permissions`

File: .github/workflows/unit-tests-code.yaml [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+      - 'go.sum'
+
+jobs:
+  test:
+    name: Unit Tests (Go)
+    runs-on: ubuntu-22.04-32core
+
+    permissions:
+      contents: read
+      id-token: write
```
Comment (reedloden) [inline .github/workflows/unit-tests-code.yaml:21]: Is this needed?

File: .github/workflows/unit-tests-helm-bypass.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+name: Unit Tests (Helm)
+run-name: Unit Tests (Helm) - ${{ github.run_id }} - @${{ github.actor }}
+
+on:
+  pull_request:
+    paths-ignore:
+      - 'examples/chart/**'
+
+jobs:
+  test:
```
Comment (reedloden) [inline .github/workflows/unit-tests-helm-bypass.yaml:11]: This is missing `permissions`

File: .github/workflows/unit-tests-helm.yaml [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    paths:
+      - 'examples/chart/**'
+
+jobs:
+  test:
```
Comment (reedloden) [inline .github/workflows/unit-tests-helm.yaml:13]: This is missing `permissions`

File: .github/workflows/unit-tests-operator-bypass.yaml [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  pull_request:
+    paths-ignore:
+      - /go.mod
+      - /go.sum
+      - operator/**
+      - api/types/**
+      - lib/tbot/**
+
+jobs:
+  test:
```
Comment (reedloden) [inline .github/workflows/unit-tests-operator-bypass.yaml:14]: This is missing `permissions`

File: .github/workflows/unit-tests-operator.yaml [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  pull_request:
+    paths:
+      - /go.mod
+      - /go.sum
+      - operator/**
+      - api/types/**
+      - lib/tbot/**
+
+jobs:
+  test:
```
Comment (reedloden) [inline .github/workflows/unit-tests-operator.yaml:17]: This is missing `permissions`

File: .github/workflows/unit-tests-rust-bypass.yaml [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+on:
+  pull_request:
+    paths-ignore:
+      - '**.rs'
+      - 'Cargo.toml'
+      - 'Cargo.lock'
+
+jobs:
+  test:
```
Comment (reedloden) [inline .github/workflows/unit-tests-rust-bypass.yaml:12]: This is missing `permissions`

File: .github/workflows/unit-tests-rust.yaml [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    branches:
+      - master
+  pull_request:
+    paths:
+      - '**.rs'
+      - 'Cargo.toml'
+      - 'Cargo.lock'
+
+jobs:
+  test:
```
Comment (reedloden) [inline .github/workflows/unit-tests-rust.yaml:15]: This is missing `permissions`

**Pull request merged**